### PR TITLE
Use FP32 scaling factors in ReduceTensor, regardless of the tensor data type (SWDEV-284921).

### PR DESCRIPTION
**Pull request by qianfengz**
Previous implementation of miopenReduceTensor() assumes alpha/beta has the same data type as the input tensor.  This change assumes/interprets alpha/beta has float type.  The miopenReduceTensor() API does not change.  

<details><summary><b><i>The change has gone through the following tests:</i></b></summary>

Regular MIOpen tests:
```bash
$ ./bin/test_reduce_test --all --float
$ ./bin/test_reduce_test --all --half
```

Manual testing:
```bash
#!/bin/bash

## for fp32
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 4 -N 0 -I 1
##
bin/MIOpenDriver reduce -D 64,3,280,81  -R 1 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 1 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 1 -O 4 -N 0 -I 1
##
bin/MIOpenDriver reduce -D 64,3,280,81  -R 2 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 2 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 2 -O 4 -N 0 -I 1

bin/MIOpenDriver reduce -D 64,3,280,81  -R 3 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 3 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 3 -O 4 -N 0 -I 1
##
bin/MIOpenDriver reduce -D 64,3,280,81  -R 2,3 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 2,3 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 2,3 -O 4 -N 0 -I 1

## for fp16
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 0 -O 2 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 0 -O 3 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 0 -O 4 -C 0 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 1 -O 2 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 1 -O 3 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 1 -O 4 -C 0 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 2 -O 2 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 2 -O 3 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 2 -O 4 -C 0 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 3 -O 2 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 3 -O 3 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 3 -O 4 -C 0 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 2,3 -O 2 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 2,3 -O 3 -C 0 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 64,3,280,81  -R 2,3 -O 4 -C 0 -N 0 -I 1

## for very small tensors
bin/MIOpenDriver reducefp16 -D 4  -R 0 -O 3 -C 0 
bin/MIOpenDriver reducefp16 -D 4  -R 0 -O 3 -C 1 
bin/MIOpenDriver reduce -D 4  -R 0 -O 3 -C 0 
bin/MIOpenDriver reduce -D 4  -R 0 -O 3 -C 1 
```
</details>

### CODE REVIEW

**Comment by atamazov** [comment]
> Previous implementation of miopenReduceTensor() assumes alpha/beta has the same data type as the input tensor. 

[Clarification] All other primitives use `float` alpha/beta (unless a bug in the library). This PR makes the Reduce Tensor primitive uniform with them.

> The miopenReduceTensor() API does not change.

[Clarification] Formally speaking, yes. Bun in reality this changes the API. More correctly, this ___fixes___ API, so this PR does not introduce _API changes_ anyway.

File: test/reduce_test.cpp [hunk 1]
```diff
@@ -317,7 +317,8 @@ struct verify_reduce_with_indices
 
                 // scale the prior dst value and add it to the accumulated value
                 if(!float_equal_zero(beta))
-                    accuVal += convert_type<compType>(output.data[dst_offset] * beta);
+                    accuVal += convert_type<compType>(output.data[dst_offset]) *
```
Comment (atamazov) [inline test/reduce_test.cpp:320]: [Q] Do our kernels perform scaling in the tensor data type (`T`) domain as well?

**Review by atamazov — APPROVED** [review]
LGTM!

File: test/reduce_test.cpp (hunk 1 above)
Comment (qianfengz) [reply test/reduce_test.cpp:320]: The alpha and beta are passed to kernel entrance both as float values,  and them alpha is converted to be same as srcType, beta is converted to be same as dstType for each specific.

**Comment by junliume** [comment]
> > Previous implementation of miopenReduceTensor() assumes alpha/beta has the same data type as the input tensor.
> 
> [Clarification] All other primitives use `float` alpha/beta (unless a bug in the library). This PR makes the Reduce Tensor primitive uniform with them.
> 
> > The miopenReduceTensor() API does not change.
> 
> [Clarification] Formally speaking, yes. Bun in reality this changes the API. More correctly, this _**fixes**_ the API, so this PR does not introduce _API changes_ anyway.

@qianfengz  Thanks for the quick actions! We do have received inquires whether this API change will affect other internal product, and let's see if this branch can fix the P1 issue.

File: src/reducetensor.cpp [hunk 2]
```diff
@@ -592,14 +590,11 @@ void ReduceTensorDescriptor::ReduceTensor(const Handle& handle,
     const std::vector<size_t> vgd_1 = {
         static_cast<size_t>(gridSize * blockSize), size_t{1}, size_t{1}};
 
-    visit_float(srcDataType, [&](auto as_float) {
-        float alphaVal = convert_type<float>(*as_float(alpha));
-        float betaVal  = convert_type<float>(*as_float(beta));
+    float alphaVal = *reinterpret_cast<const float*>(alpha);
+    float betaVal  = *reinterpret_cast<const float*>(beta);
```
Comment (atamazov) [inline src/reducetensor.cpp:594]: :anchor:

File: src/reducetensor.cpp [hunk 3]
```diff
@@ -614,14 +609,9 @@ void ReduceTensorDescriptor::ReduceTensor(const Handle& handle,
...
-            float alphaVal = convert_type<float>(*as_float(alpha));
-            float betaVal  = convert_type<float>(*as_float(beta));
-
-            handle.AddKernel(
-                algo_name, network_config, program_name, kernel_name2, vld_2, vgd_2, param)(
-                alphaVal, A, betaVal, C, ws_buf1_global, ws_buf2_bytes_offset, indices);
-        });
+        handle.AddKernel(
+            algo_name, network_config, program_name, kernel_name2, vld_2, vgd_2, param)(
+            alphaVal, A, betaVal, C, ws_buf1_global, ws_buf2_bytes_offset, indices);
```
Comment (atamazov) [inline src/reducetensor.cpp:614]: :anchor:

**Pull request merged**